ecs-agent/acs/session/payload_responder.go

```diff
@@ -0,0 +1,162 @@
+	respond            wsclient.RespondFunc
+}
+
+// NewPayloadResponder returns an instance of the payloadResponder struct.
+func NewPayloadResponder(taskMessageHandler TaskMessageHandler,
+	responseSender wsclient.RespondFunc) wsclient.RequestResponder {
+	r := &payloadResponder{
+		taskMessageHandler: taskMessageHandler,
+	}
+	r.respond = ResponseToACSSender(r.Name(), responseSender)
+	return r
+}
+
+func (*payloadResponder) Name() string { return "payload responder" }
+
+func (r *payloadResponder) HandlerFunc() wsclient.RequestHandler {
+	return r.handlePayloadMessage
+}
+
+func (r *payloadResponder) handlePayloadMessage(message *ecsacs.PayloadMessage) {
+	logger.Debug(fmt.Sprintf("Handling %s", PayloadMessageName))
+
+	// Validate fields in the message.
+	if err := validatePayloadMessage(message); err != nil {
+		logger.Error(fmt.Sprintf("Error validating %s received from ECS", PayloadMessageName), logger.Fields{
+			field.Error: err,
+		})
+		return
+	}
+
+	// Handle payload message.
+	err := r.taskMessageHandler.ProcessTaskMessage(message, r.ackFunc)
+	if err != nil {
+		logger.Critical(fmt.Sprintf("Unable to handle %s", PayloadMessageName), logger.Fields{
+			field.MessageID: aws.StringValue(message.MessageId),
+			field.Error:     err,
+		})
+	}
+}
+
+// ackFunc sends ACKs of the payload message and of the credentials associated with the tasks contained in the payload
+// message.
+func (r *payloadResponder) ackFunc(payloadAck *ecsacs.AckRequest, credsAcks []*ecsacs.IAMRoleCredentialsAckRequest) {
+	go r.sendAck(payloadAck)
+
+	for _, credsAck := range credsAcks {
+		go r.sendAck(credsAck)
+	}
+}
+
+// sendAck handles the sending of an individual specific ACK, assuming it is of type
+// ecsacs.IAMRoleCredentialsAckRequest or ecsacs.AckRequest.
+func (r *payloadResponder) sendAck(ackRequest interface{}) {
+	var credentialsAck *ecsacs.IAMRoleCredentialsAckRequest
+	var payloadMessageAck *ecsacs.AckRequest
+	credentialsAck, ok := ackRequest.(*ecsacs.IAMRoleCredentialsAckRequest)
+	if ok {
+		logger.Debug(fmt.Sprintf("ACKing credentials associated with %s", PayloadMessageName), logger.Fields{
+			"credentialsID": aws.StringValue(credentialsAck.CredentialsId),
+			field.MessageID: aws.StringValue(credentialsAck.MessageId),
+		})
+	} else {
+		payloadMessageAck, ok = ackRequest.(*ecsacs.AckRequest)
+		if ok {
+			logger.Debug(fmt.Sprintf("ACKing %s", PayloadMessageName), logger.Fields{
+				field.MessageID: aws.StringValue(payloadMessageAck.MessageId),
+			})
+		} else {
+			logger.Error(fmt.Sprintf("Error sending acknowledgement: %s",
+				"ackRequest does not hold type ecsacs.IAMRoleCredentialsAckRequest or ecsacs.AckRequest"))
+			return
+		}
+	}
+
+	err := r.respond(ackRequest)
+
+	if err != nil {
+		if credentialsAck != nil {
+			logger.Warn(fmt.Sprintf("Error acknowledging credentials associated with %s",
+				PayloadMessageName), logger.Fields{
+				"credentialsID": aws.StringValue(credentialsAck.CredentialsId),
```

Comment:
I am not sure if it helps much to log credentials ID since it generally does not mean anything and is a uuid. Is there any reason we decide to add this in the log?

Reply:
Credentials ID are currently already being logged in the existing code for this case, so I decided to preserve this: https://github.com/aws/amazon-ecs-agent/blob/dev/agent/acs/handler/payload_handler.go#L200-L201

The payload message [contains a list of tasks](https://github.com/aws/amazon-ecs-agent/blob/dev/ecs-agent/acs/model/ecsacs/api.go#L1389) which each can contain unique credentials (that have a unique credentials ID). I think logging credentials ID could be useful here in the event a payload message contains multiple tasks which require different credentials, and that ACKing of some credentials fails but some are successful. That way, we can differentiate which credentials were successfully ACKed and which were not (only logging the message ID would make it unclear which credentials in such a payload message were not able to be ACKed).

Reply:
ack.